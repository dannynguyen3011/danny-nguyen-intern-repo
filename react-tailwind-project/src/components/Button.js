import React from 'react';
import PropTypes from 'prop-types';

/**
 * Reusable Button component with Tailwind CSS styling
 * Supports different variants, sizes, and states
 */
const Button = ({
  children,
  onClick,
  variant = 'primary',
  size = 'medium',
  disabled = false,
  fullWidth = false,
  type = 'button',
  className = '',
  ...props
}) => {
  // Base button styles
  const baseStyles = 'font-semibold rounded-lg transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 transform active:scale-95 disabled:opacity-50 disabled:cursor-not-allowed disabled:transform-none';

  // Variant styles
  const variants = {
    primary: 'bg-blue-600 text-white hover:bg-blue-700 active:bg-blue-800 focus:ring-blue-500 shadow-lg hover:shadow-xl',
    secondary: 'bg-gray-600 text-white hover:bg-gray-700 active:bg-gray-800 focus:ring-gray-500 shadow-lg hover:shadow-xl',
    success: 'bg-green-600 text-white hover:bg-green-700 active:bg-green-800 focus:ring-green-500 shadow-lg hover:shadow-xl',
    danger: 'bg-red-600 text-white hover:bg-red-700 active:bg-red-800 focus:ring-red-500 shadow-lg hover:shadow-xl',
    warning: 'bg-yellow-600 text-white hover:bg-yellow-700 active:bg-yellow-800 focus:ring-yellow-500 shadow-lg hover:shadow-xl',
    info: 'bg-cyan-600 text-white hover:bg-cyan-700 active:bg-cyan-800 focus:ring-cyan-500 shadow-lg hover:shadow-xl',
    outline: 'bg-transparent border-2 border-blue-600 text-blue-600 hover:bg-blue-600 hover:text-white active:bg-blue-700 focus:ring-blue-500',
    ghost: 'bg-transparent text-blue-600 hover:bg-blue-50 active:bg-blue-100 focus:ring-blue-500',
    gradient: 'bg-gradient-to-r from-purple-600 to-blue-600 text-white hover:from-purple-700 hover:to-blue-700 active:from-purple-800 active:to-blue-800 focus:ring-purple-500 shadow-lg hover:shadow-xl',
  };

  // Size styles
  const sizes = {
    small: 'px-3 py-1.5 text-sm',
    medium: 'px-4 py-2 text-base',
    large: 'px-6 py-3 text-lg',
    extraLarge: 'px-8 py-4 text-xl',
  };

  // Width styles
  const widthStyles = fullWidth ? 'w-full' : '';

  // Combine all styles
  const buttonClasses = `
    ${baseStyles}
    ${variants[variant]}
    ${sizes[size]}
    ${widthStyles}
    ${className}
  `.trim().replace(/\s+/g, ' ');

  return (
    <button
      type={type}
      onClick={onClick}
      disabled={disabled}
      className={buttonClasses}
      {...props}
    >
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.node.isRequired,
  onClick: PropTypes.func,
  variant: PropTypes.oneOf([
    'primary', 'secondary', 'success', 'danger', 
    'warning', 'info', 'outline', 'ghost', 'gradient'
  ]),
  size: PropTypes.oneOf(['small', 'medium', 'large', 'extraLarge']),
  disabled: PropTypes.bool,
  fullWidth: PropTypes.bool,
  type: PropTypes.oneOf(['button', 'submit', 'reset']),
  className: PropTypes.string,
};

export default Button;
