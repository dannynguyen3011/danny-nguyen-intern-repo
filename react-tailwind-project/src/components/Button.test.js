/**
 * Tests for the reusable Button component
 * Demonstrates testing component variants, props, and user interactions
 */

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import Button from './Button';

describe('Button Component', () => {

  // Basic rendering tests
  describe('rendering', () => {
    test('renders button with children', () => {
      render(<Button>Click me</Button>);
      
      expect(screen.getByRole('button')).toBeInTheDocument();
      expect(screen.getByText('Click me')).toBeInTheDocument();
    });

    test('renders with default props', () => {
      render(<Button>Default Button</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('type', 'button');
      expect(button).not.toBeDisabled();
    });
  });

  // Variant testing
  describe('variants', () => {
    const variants = [
      'primary', 'secondary', 'success', 'danger', 
      'warning', 'info', 'outline', 'ghost', 'gradient'
    ];

    variants.forEach(variant => {
      test(`renders ${variant} variant correctly`, () => {
        render(<Button variant={variant}>Test Button</Button>);
        
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
        expect(button).toHaveTextContent('Test Button');
      });
    });

    test('applies correct CSS classes for primary variant', () => {
      render(<Button variant="primary">Primary</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('bg-blue-600');
      expect(button).toHaveClass('text-white');
      expect(button).toHaveClass('hover:bg-blue-700');
    });

    test('applies correct CSS classes for outline variant', () => {
      render(<Button variant="outline">Outline</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('bg-transparent');
      expect(button).toHaveClass('border-2');
      expect(button).toHaveClass('border-blue-600');
    });
  });

  // Size testing
  describe('sizes', () => {
    const sizes = ['small', 'medium', 'large', 'extraLarge'];

    sizes.forEach(size => {
      test(`renders ${size} size correctly`, () => {
        render(<Button size={size}>Size Test</Button>);
        
        const button = screen.getByRole('button');
        expect(button).toBeInTheDocument();
        expect(button).toHaveTextContent('Size Test');
      });
    });

    test('applies correct CSS classes for small size', () => {
      render(<Button size="small">Small</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('px-3');
      expect(button).toHaveClass('py-1.5');
      expect(button).toHaveClass('text-sm');
    });

    test('applies correct CSS classes for large size', () => {
      render(<Button size="large">Large</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('px-6');
      expect(button).toHaveClass('py-3');
      expect(button).toHaveClass('text-lg');
    });
  });

  // Props testing
  describe('props', () => {
    test('handles onClick prop', async () => {
      const handleClick = jest.fn();
      
      render(<Button onClick={handleClick}>Clickable</Button>);
      
      const button = screen.getByRole('button');
      await userEvent.click(button);
      
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    test('handles disabled state', () => {
      const handleClick = jest.fn();
      
      render(<Button disabled onClick={handleClick}>Disabled</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toBeDisabled();
      
      fireEvent.click(button);
      expect(handleClick).not.toHaveBeenCalled();
    });

    test('applies fullWidth prop', () => {
      render(<Button fullWidth>Full Width</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('w-full');
    });

    test('handles different button types', () => {
      const { rerender } = render(<Button type="submit">Submit</Button>);
      expect(screen.getByRole('button')).toHaveAttribute('type', 'submit');
      
      rerender(<Button type="reset">Reset</Button>);
      expect(screen.getByRole('button')).toHaveAttribute('type', 'reset');
    });

    test('applies custom className', () => {
      render(<Button className="custom-class">Custom</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('custom-class');
    });

    test('passes through additional props', () => {
      render(<Button data-testid="custom-button" aria-label="Custom label">Test</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveAttribute('data-testid', 'custom-button');
      expect(button).toHaveAttribute('aria-label', 'Custom label');
    });
  });

  // Accessibility tests
  describe('accessibility', () => {
    test('has button role', () => {
      render(<Button>Accessible Button</Button>);
      
      expect(screen.getByRole('button')).toBeInTheDocument();
    });

    test('is keyboard accessible', async () => {
      const handleClick = jest.fn();
      
      render(<Button onClick={handleClick}>Keyboard Test</Button>);
      
      const button = screen.getByRole('button');
      button.focus();
      
      expect(button).toHaveFocus();
      
      await userEvent.keyboard('{Enter}');
      expect(handleClick).toHaveBeenCalledTimes(1);
      
      await userEvent.keyboard(' ');
      expect(handleClick).toHaveBeenCalledTimes(2);
    });

    test('applies focus styles', () => {
      render(<Button>Focus Test</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('focus:outline-none');
      expect(button).toHaveClass('focus:ring-2');
    });

    test('shows disabled state visually', () => {
      render(<Button disabled>Disabled Button</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toHaveClass('disabled:opacity-50');
      expect(button).toHaveClass('disabled:cursor-not-allowed');
    });
  });

  // User interaction tests
  describe('user interactions', () => {
    test('handles multiple clicks', async () => {
      const handleClick = jest.fn();
      
      render(<Button onClick={handleClick}>Multi Click</Button>);
      
      const button = screen.getByRole('button');
      
      await userEvent.click(button);
      await userEvent.click(button);
      await userEvent.click(button);
      
      expect(handleClick).toHaveBeenCalledTimes(3);
    });

    test('prevents clicks when disabled', async () => {
      const handleClick = jest.fn();
      
      render(<Button disabled onClick={handleClick}>Disabled Click</Button>);
      
      const button = screen.getByRole('button');
      
      await userEvent.click(button);
      
      expect(handleClick).not.toHaveBeenCalled();
    });

    test('handles mouse events', async () => {
      render(<Button>Hover Test</Button>);
      
      const button = screen.getByRole('button');
      
      await userEvent.hover(button);
      expect(button).toHaveClass('hover:bg-blue-700');
      
      await userEvent.unhover(button);
      // Button should still have hover class (CSS handles the state)
    });
  });

  // Edge cases
  describe('edge cases', () => {
    test('handles empty children', () => {
      render(<Button></Button>);
      
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
      expect(button).toHaveTextContent('');
    });

    test('handles complex children', () => {
      render(
        <Button>
          <span>Icon</span>
          <span>Text</span>
        </Button>
      );
      
      const button = screen.getByRole('button');
      expect(button).toHaveTextContent('IconText');
      expect(screen.getByText('Icon')).toBeInTheDocument();
      expect(screen.getByText('Text')).toBeInTheDocument();
    });

    test('handles invalid variant gracefully', () => {
      // Should not crash with invalid variant
      render(<Button variant="invalid">Invalid Variant</Button>);
      
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
    });
  });

  // Performance tests
  describe('performance', () => {
    test('renders quickly', () => {
      const start = performance.now();
      render(<Button>Performance Test</Button>);
      const end = performance.now();
      
      expect(end - start).toBeLessThan(50);
    });

    test('handles rapid re-renders', () => {
      const { rerender } = render(<Button>Test 1</Button>);
      
      const start = performance.now();
      for (let i = 0; i < 20; i++) {
        rerender(<Button variant="primary">Test {i}</Button>);
      }
      const end = performance.now();
      
      expect(end - start).toBeLessThan(100);
    });
  });

  // Snapshot tests
  describe('snapshots', () => {
    test('matches snapshot for primary button', () => {
      const { container } = render(<Button variant="primary">Primary Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });

    test('matches snapshot for disabled button', () => {
      const { container } = render(<Button disabled>Disabled Button</Button>);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
