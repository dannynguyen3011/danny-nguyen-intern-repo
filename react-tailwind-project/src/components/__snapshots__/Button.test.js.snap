// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component snapshots matches snapshot for disabled button 1`] = `
<button
  class="font-semibold rounded-lg transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 transform active:scale-95 disabled:opacity-50 disabled:cursor-not-allowed disabled:transform-none bg-blue-600 text-white hover:bg-blue-700 active:bg-blue-800 focus:ring-blue-500 shadow-lg hover:shadow-xl px-4 py-2 text-base"
  disabled=""
  type="button"
>
  Disabled Button
</button>
`;

exports[`Button Component snapshots matches snapshot for primary button 1`] = `
<button
  class="font-semibold rounded-lg transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 transform active:scale-95 disabled:opacity-50 disabled:cursor-not-allowed disabled:transform-none bg-blue-600 text-white hover:bg-blue-700 active:bg-blue-800 focus:ring-blue-500 shadow-lg hover:shadow-xl px-4 py-2 text-base"
  type="button"
>
  Primary Button
</button>
`;
